ea-2021-04-ndvi-automation-brian-pinke.py

```diff
@@ -0,0 +1,790 @@
+
+
+# # Figure 1: Plot 1 - Mean NDVI For Each Site Across the Year (50 points)
+# 
+# Create a plot of the mean normalized difference vegetation index (NDVI) for the two different sites in the United States across the year: 
+# 
+# * NDVI on the x axis and formatted dates on the y for both NEON sites on one figure/axis object.
+# * Each site should be identified with a different color in the plot and legend.
+# * The final plot **data should be cleaned to remove the influence of clouds**.
+# * Be sure to include appropriate title and axes labels.
+# 
+# Add additional cells as needed for processing data (e.g. defining functions, etc), but be sure to:
+# * follow the instructions in the code cells that have been provided to ensure that you are able to use the sanity check tests that are provided. 
+# * include only the plot code in the cell identified for the final plot code below
+
+# ## Task 1: 
+# 
+# In the cell below, create a single dataframe containing MEAN NDVI, the site name, 
+# and the date of the data for the HARV site 
+# scene `HARV/landsat-crop/LC080130302017031701T1-SC20181023151837`.  The column names for the  final
+# DataFrame should be`mean_ndvi`, and `site`, and the data should be **indexed on the date**. 
+# 
+# Use the functions that we reviewed in class (or create your own versions of them) to implement your code
+# 
+# ### In the Cell below Place  All Functions Needed to Run this Notebook (20 points)
+
+# In[4]:
+
+
+### DO NOT REMOVE THIS LINE OR EDIT / MOVE THIS CELL ###
+start_time = datetime.now()
+
+
+# # Functions to process and optimize code
+# * These four functions are created to reduce copypasta code throughout the notebook
+# * The use of functions also allows for easier modification of code and flow paths
+# * Open_clean_bands function utilizes 'from_disk' to slice where it's only grabbing the data that is needed, saving time and memory resources
+
+# In[5]:
+
+
+# In this cell place all of the functions needed to run your notebook
+# You will be graded here on function application, docstrings, efficiency so ensure
+# All functions are placed here!
+
+# YOUR CODE HERE
+def open_clean_bands(band_path,
+                     crop_bound,
+                     valid_range=None,
+                     a_mask=None,
+                     vals=None):
+    """Open and mask a single landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    band_path : string
+        A path to the array to be opened
+    crop_bound : geopandas GeoDataFrame
+        A geopandas dataframe to be used to crop the raster data using 
+            rasterio mask().
+    valid_range : tuple (optional)
+        A tuple of min and max range of values for the data. Default = None
+    a_mask : xarray DataArray
+        An xarray DataArray with values that have not yet been set to 1
+    vals : list
+        A list of values needed to create the cloud mask
+
+    Returns
+    -----------
+    arr : xarray DataArray
```

Comment:
I always get confused about whether the doc string should include the explicit name of the returned variable or if that matters. In this case should the Returns be "band_crop" instead of "arr".

Reply:
ahhh yup you're totally right! Good catch!